### ISSUE TITLE
Prevent mouse-drag from activating visual-state (fix selection behavior in pdf-tools)

### ISSUE THREAD

**Issue by dalanicolai** [emacs-evil/evil#1671]
I'm opening this issue due to the unsatisfactory selection behavior in pdf-tools when using evil. Without the evil package, one can just select text in pdf-tools at any time. However, when using evil, then after some selection has been made, evil switches to `visual-state` in which it is impossible to make another selection (one first has to exit visual-state). As having a separate visual-state in pdf-tools is more or less useless, it would be great if in this situation evil could be prevented from entering visual-state. Then we can just bind `evil-yank` etc. in the normal/visual-mode-map directly.

Of course, we could just use and redefine bindings in `emacs-state` in pdf-tools, but in case of Spacemacs/Doom-Emacs, we would have to redefine the keymap behind the `SPC`, and define our evil keybindings within `emacs-state` making it impossible to switch between emacs/evil keybindings. Therefore, the proposed solution is preferable.

I have found the answer [here](https://github.com/emacs-evil/evil/issues/936#issuecomment-335043309), but it is unsatisfactory, because obviously, even without `evil-mouse-drag-region` being used, it is still evil that is responsible for the switching to visual-state (vanilla Emacs does not switch (evil-)state after dragging).

So the request is, could some configuration variable/extra function get implemented to prevent evil from entering 'visual-state' after a mouse-drag (or is it already available)? Ideally, one either would define a list with modes where one wants evil to be prevented from entering `visual-state` after mouse drag, or one would bind an alternative function to `evil-mouse-drag-region` in `motion-state`, which would prevent evil from entering `visual-state`.

# Issue type
- Enhancement request
- Question

# Reproduction steps
- Start Emacs and activate evil
- do `M-:` `(define-key evil-motion-state-map [down-mouse-1] nil)` to unbind `evil-mouse-drag-region` in motion-state
- then back in the 'welcome/home' buffer (where evil uses `motion-state`), select text by dragging using the mouse

For the pdf-tools part:
- open pdf-tools in evil-state
- select some text
- immediately try to select text again (it is not possible because evil, in this case unnecessarily, switches to visual state)
- switch to emacs-state (`C-z`) and repeat, small but huge difference :)

# Expected behavior
We get the expected behavior, however, we would like in some situations to prevent evil from entering `visual-state`

Thanks!

**Comment by dalanicolai** [comment]
I have found that I can just remove `evil-visual-activate-hook` from the `activate-mark-hook`, so that work perfectly fine. Anyway, thanks for evil!

**Comment by dalanicolai** [comment]
I reopen the issue, because actually I think it is better to let you decide about the best solution. Anyway the fix in https://github.com/syl20bnr/spacemacs/pull/15740/files (which is a fix of https://github.com/syl20bnr/spacemacs/pull/15736) seems to work well.

**Comment by tomdl89** [comment]
@dalanicolai what does the instructions `open pdf-tools in evil-state` mean? I am not a spacemacs user, so can I have exact instructions to reproduce this with only evil and pdf-tools modes loaded? If a line or two of elisp is required to get to this problem, that's fine, but if it's only reachable from spacemacs doing something strange, I'd expect the fix you've merged should suffice?

**Comment by dalanicolai** [comment]
@tomdl89 Ah sorry, that is sloppy language indeed. I mean, open some pdf with pdf-tools while using evil. The buffer will open in `normal-state`. Now select some text (i.e. using the mouse), and immediately try to again select some other text... it does not work because evil has switched to visual state.

Now try the same, i.e. try making a selection twice, while having switched to emacs-state (using C-z)... it simply selects text both times.

As visual state is not useful in pdf-tools, it would be nice to prevent evil from switching to visual-state when using evil (normal- or motion-state).

**Comment by tomdl89** [comment]
In my emacs, when I open a PDF, it opens with DocView. I do `M-x pdf-tools-view` which gives me the pdf-tools interface. But doesn't put me in `normal-state`. So I don't see these problems. I'm guessing either spacemacs has a hook which enables evil mode, or I'm doing the wrong thing by doing `M-x pdf-tools-view`.

**Comment by dalanicolai** [comment]
Yes Spacemacs opens modes derived from `special-mode` in `motion-state` by adding `special-mode` to the list of `evil-motion-state-modes`.

However, I have a somewhat 'vanilla' Emacs configuration also. I guess you should run `M-x pdf-tools-install` before opening the pdf, so it will open directly in `pdf-view-mode` (I guess in `normal-state`).

Sorry, I somehow expected most/all Emacs users were using pdf-tools already (therefore I did not think about providing installation instructions).

**Comment by tomdl89** [comment]
And do you actually want to be in evil-motion-state in pdf-tools mode?

**Comment by dalanicolai** [comment]
Yes (I have edited my previous comment, I was accidentally testing in Spacemacs, my vanilla config opens in `motion-state`)

**Comment by dalanicolai** [comment]
B.t.w. I now realize, that [evil-collection](https://github.com/emacs-evil/evil-collection) does not use `motion-state` but instead prevents `normal-state` from entering `visual-state` (for the relevant cases). I will see if I can find out how they achieve that behavior.

**Comment by tomdl89** [comment]
I noticed they have this binding:
https://github.com/emacs-evil/evil-collection/blob/6306aa78a94bc517b77bdecde830802c4ac1d9bd/modes/pdf/evil-collection-pdf.el#L170
in `'pdf-view-mode-map` which may help? I don't use evil-collection either, btw.

**Comment by dalanicolai** [comment]
I am more or less sure that that binding does not help, as I think it is that `set-region` function which makes evil switch to `visual-state`. But I can quickly check it, and then I will update this comment...

**Comment by tomdl89** [comment]
I just executed
```
(evil-define-key 'normal pdf-view-mode-map (kbd "<down-mouse-1>") 'pdf-view-mouse-set-region)
```
before entering evil-motion-state, and it seems to have fixed the problem.

**Comment by dalanicolai** [comment]
Mmm. very confusing, now pdf-tools in my vanilla config does open in `normal-state` (while it should open in `motion-state`).

Anyway, using the `define-key` from your previous command does not solve the issue (which is to be expected, as it calls `pdf-view-mouse-set-region`, which activate a region, making evil switch to `visual-state`).

**Comment by tomdl89** [comment]
I don't think activating a region in itself makes evil switch to visual-state, but I may be wrong. Looks like it needs some more investigation. Feel free to do so - I'm not sure I'll get the chance any time soon.

**Comment by dalanicolai** [comment]
I am almost sure it does (because I have investigated that before). Are you sure that you can reproduce that fix by doing the`define-key` first?

**Comment by dalanicolai** [comment]
Thanks again, I will investigate a little more. Don't spend more time on it. And thanks for your time already. I will keep you updated...

**Comment by dalanicolai** [comment]
Okay, I have found a way to fix this issue.

We could simply add the following `unless` in `evil-local-mode`:
```
    (unless (eq major-mode 'pdf-view-mode)
      (add-hook 'activate-mark-hook 'evil-visual-activate-hook nil t))
```

This seems to have the desired effect without unwanted side effects.

I can imagine that `pdf-view-mode` is the only mode where this behavior is desired, so we could just add this change as posted here.
However, I could also imagine that you would prefer to add some `no-visual-state-modes` variable, which makes it more general and provides the possibility to better document it.

**Comment by tomdl89** [comment]
Great work! Thanks. Yeah I think I prefer the option of having an `evil-no-visual-state-modes` variable, but I think it's fine to pre-populate it with one item: `'pdf-view-mode`. There may be other modes we don't know about that could benefit from this.

**Comment by workcomplete** [comment]
Glad to see a fix for this! I can confirm that this is an issue in other modes that allow for mouse use, like dictionary-el

**Comment by dalanicolai** [comment]
I suddenly got another, embarrassingly simple, idea to fix this. And it comes close to what you suggested [here](https://github.com/emacs-evil/evil/issues/1671#issuecomment-1268655376)...

Obviously, we can just try to add those mouse bindings to the visual state map, which seems to work perfectly fine.
I guess, that is probably, the 'cleanest' solution. So I will simply add those bindings for Spacemacs. And maintainers of other packages/distributions can do the same.

@tomdl89 Thanks again for your time and effort. I wish I had realized this earlier. I will close this issue for now, and I hope for forever, then...

**Comment by tomdl89** [comment]
I still think the variable idea may be a good idea for those not using distros 🤔 If not, clear step by step eli5 instructions left as a comment here would be good for posterity, I think.

**Comment by dalanicolai** [comment]
Well, the problem with the 'variable' option, is that I don't know how evil works, and although the solution seems to work fine, I am not sure if the fix would be an 'elegant' thing to do/logical way to achieve it.

But yeah, I should have mentioned this issue in [the new PR](https://github.com/syl20bnr/spacemacs/pull/15762).

So the steps are simply to add the default pdf-view-mode mouse bindings, to the visual state as follows:
```lisp
(evil-define-key 'visual pdf-view-mode-map
  "y" 'pdf-view-kill-ring-save
  (kbd "<C-down-mouse-1>") 'pdf-view-mouse-extend-region
  (kbd "<M-down-mouse-1>") 'pdf-view-mouse-set-region-rectangle
  (kbd "<down-mouse-1>")  'pdf-view-mouse-set-region)
```
